apps/game_backend/lib/game_backend/users/currencies.ex

```diff
@@ -85,16 +83,33 @@ defmodule GameBackend.Users.Currencies do
   """
   def add_currency(user_id, currency_id, amount) do
     result =
-      with %UserCurrency{} = user_currency <- get_user_currency(user_id, currency_id),
-           changeset <-
-             UserCurrency.update_changeset(user_currency, %{
-               amount: max(user_currency.amount + amount, 0)
-             }) do
-        Repo.update(changeset)
-      else
+      case get_user_currency(user_id, currency_id) do
+        %UserCurrency{} = user_currency ->
+          new_amount =
+            case get_user_currency_cap(user_id, currency_id) do
+              %UserCurrencyCap{cap: cap} ->
+                if cap <= user_currency.amount do
+                  # Cap reached, don't add anything.
+                  user_currency.amount
+                else
+                  # Cap not reached, add the amount. We are fine with allowing overflows.
+                  user_currency.amount + amount
+                end
+
+              nil ->
+                # No cap, just add the amount.
+                user_currency.amount + amount
+            end
+            # We don't want users with negative currencies
+            |> max(0)
+
+          user_currency
+          |> UserCurrency.update_changeset(%{amount: new_amount})
+          |> Repo.update()
+
         nil ->
           # User has none of this currency, create it with given amount
-          insert_user_currency(%{user_id: user_id, currency_id: currency_id, amount: amount})
+          insert_user_currency(%{user_id: user_id, currency_id: currency_id, amount: max(amount, 0)})
```

Comment:
Don't we need to check the cap here?

Reply:
We don't because we won't have currency caps of zero (at least it doesn't make any sense for now, we might want to revisit it in the future) so this will in the worst case be the first overflow of the cap, which we want to allow!

Reply:
Sounds good!